Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,479 @@
+  by_cases h: i=j
+  · simp_rw [h]
+    unfold eigenRowEquiv finRankEquivEigsMulConjTranspose
+      Equiv.sumCongr Sum.map
+    simp only [ne_eq, Equiv.symm_trans_apply, Equiv.symm_symm, Equiv.coe_fn_symm_mk,
+      Sum.elim_inl, Equiv.sumCompl_apply_inl, of_apply,
+      diagonal_apply_eq, Function.comp_apply]
+  · rw [diagonal_apply_ne, of_apply, diagonal_apply_ne]
+    rw [ne_eq, EmbeddingLike.apply_eq_iff_eq, Sum.inl.injEq]
+    assumption'
+
+lemma S_block (A : Matrix (Fin M) (Fin N) 𝕂) :
+    (reindex
+      (eigenColumnEquiv A)
+      (eigenColumnEquiv A))
+        ( diagonal ( (isHermitian_transpose_mul_self A).eigenvalues)) =
+          fromBlocks A.svdμ 0 0 0 := by
+  let hz := S_zero_blocks A
+  rw [← svdS, ← fromBlocks_toBlocks (A.svdS), ← S_σpos_block, hz.1, hz.2.1, hz.2.2]
+
+lemma S'_block (A : Matrix (Fin M) (Fin N) 𝕂) :
+    (reindex
+      (eigenRowEquiv A)
+      (eigenRowEquiv A))
+        (diagonal ( (isHermitian_mul_conjTranspose_self A).eigenvalues)) =
+          fromBlocks A.svdμ' 0 0 0 := by
+  let hz := S'_zero_blocks A
+  rw [← svdS', ← fromBlocks_toBlocks (A.svdS'), ← S'_σpos_block, hz.1, hz.2.1, hz.2.2]
+
+lemma V_columns (A : Matrix (Fin M) (Fin N) 𝕂) :
+    (reindex (Equiv.refl (Fin N)) (eigenColumnEquiv A))
+      (isHermitian_transpose_mul_self A).eigenvectorMatrix =
+      fromColumns A.svdV₁ A.svdV₂ := by
+  rw [reindex_apply, fromColumns, svdV₁, svdV₂, toColumns₁, toColumns₂]
+  funext i j
+  cases' j with j j <;>
+  simp only [Equiv.refl_symm, Equiv.coe_refl, submatrix_apply, id_eq,
+    reindex_apply, of_apply, Sum.elim_inl, Sum.elim_inr]
+
+lemma reduced_spectral_theorem (A : Matrix (Fin M) (Fin N) 𝕂) :
+    Aᴴ  *  A = A.svdV₁  *  (A.svdμ.map (algebraMap ℝ 𝕂))  *  A.svdV₁ᴴ := by
+  let hAHA := isHermitian_transpose_mul_self A
+  -- "Ugly" (submatrix_mul_equiv) explicit rewrites: each one on its own line for
+  -- readability!!
+  rw [← submatrix_id_id (Aᴴ * A), IsHermitian.spectral_theorem' hAHA,
+    ← IsHermitian.conjTranspose_eigenvectorMatrix, Matrix.mul_assoc,
+    ← submatrix_mul_equiv
+      hAHA.eigenvectorMatrix (diagonal (IsROrC.ofReal ∘ hAHA.eigenvalues)  *
+      (hAHA.eigenvectorMatrixᴴ)) _ (eigenColumnEquiv A).symm _,
+    ← submatrix_mul_equiv (diagonal (IsROrC.ofReal ∘ hAHA.eigenvalues)) (hAHA.eigenvectorMatrixᴴ)
+    _ (eigenColumnEquiv A).symm _,
+    ← @IsROrC.algebraMap_eq_ofReal 𝕂]
+  simp_rw [Function.comp]
+  rw [← diagonal_map, submatrix_map, ← reindex_apply, ← Equiv.coe_refl, ← Equiv.refl_symm,
+    ← reindex_apply, ←conjTranspose_submatrix, ← reindex_apply, S_block, V_columns,
+    conjTranspose_fromColumns_eq_fromRows_conjTranspose, fromBlocks_map, fromBlocks_mul_fromRows,
+    fromColumns_mul_fromRows]
+  simp only [map_zero, Matrix.map_zero, Matrix.zero_mul, add_zero, Matrix.mul_zero]
+  rw [Matrix.mul_assoc]
+  apply map_zero
+
+lemma reduced_spectral_theorem' (A : Matrix (Fin M) (Fin N) 𝕂) :
+    A  *  Aᴴ = A.svdU₁'  *  (A.svdμ'.map (algebraMap ℝ 𝕂))  *  A.svdU₁'ᴴ := by
```

Comment:
As above:
```suggestion
/-- **Reduced spectral theorem**, left eigenvector version. -/
lemma U₁'_mul_μ'_mul_U₁'_conjTranspose (A : Matrix (Fin M) (Fin N) 𝕂) :
    A.svdU₁'  *  (A.svdμ'.map (algebraMap ℝ 𝕂))  *  A.svdU₁'ᴴ = A  *  Aᴴ := by
```

Reply:
If you want to keep `U₁'` around (see the comment on `U₁'`), could you also add the `U₁` (with no `'`) version?

Reply:
OK I added a version for U1 (without the dash/prime). It is called `U₁_mul_μ_mul_U₁_conjTranspose`